File: examples/Boids_Shm/src/BirdShm.cpp
#include "BirdShm.h"
#include "SyncConTMtbb.h"

#include <iostream>
#include <math.h>
#include <time.h>
#include <string>

using namespace std;
using namespace boost::random;

extern const double Length;
extern const double Width;
extern const double Height;

void BirdShm::Init()
{
	acceleration = std::move( Vector3D(0, 0, 0) );
    
    maxSpeed = 3.5;
    maxForce = 0.5;

	seperate_range = 20;
	alignment_range = 50;
	cohesion_range = 50;
	delta = 1;
	
	//!< 将数据写入LP共享内存区
	BirdTrace bt(EntityID(), position, velocity);
	dynamic_cast<BoidsState*>(getIntraLPState())->add(EntityID(), bt);

	startTick(delta);

}

void BirdShm::execute(SimMsg *pmsg)
{
	int ev_type = pmsg->getMsgType();

	switch (ev_type)
	{
		case SIM_ENT_TICKMSG:
		{
			CycleWork();

			//!< 将数据写入LP共享内存区
			BirdTrace bt(EntityID(), position, velocity);
			dynamic_cast<BoidsState*>(getIntraLPState())->modify(EntityID(), bt);

			break;
		}
		default: {
			SIMDBG(0, "cannot find the corresponding handler! " << ev_type );
			break;
		}
	}
}

void BirdShm::Terminate(SimTime ts)
{
	endTick(delta);	
}

void BirdShm::CycleWork()
{
	flock();

    //To make the slow down not as abrupt
    acceleration *= 0.4;

    // Update velocity
    velocity += acceleration;

    // Limit speed
    velocity.limit(maxSpeed);
	position += velocity*delta;

    // Reset accelertion to 0 each cycle
    acceleration *= 0.0;

	//!< 越界时修正位置
	if (position.x < 0)    position.x += Length;
    if (position.y < 0)    position.y += Width;
	if (position.z < 0)    position.z += Height;
    if (position.x > Length) position.x -= Length;
    if (position.y > Width) position.y -= Width;
	if (position.z > Height) position.z -= Height;

	//BirdTrace* birdt = new BirdTrace(id, position, velocity);
	//SimMsg* posmsg = new SimMsg(T_BirdTrace, birdt);
	//post("BirdTrace", posmsg, 0.1);

	//neighborbirds.clear();
}

void BirdShm::recvBirdTrace(BirdTrace* stc)
{
	// neighborbirds.push_back(stc);
}

// 决定bird的移动方向
void BirdShm::flock()
{
    Vector3D sep, ali, coh;
    for( auto& neighborbirds : getSharedState()->getReadSharedState())
    {
        sep += Separation(dynamic_cast<BoidsState*>(neighborbirds) );
        ali += Alignment(dynamic_cast<BoidsState*>(neighborbirds));
        coh += Cohesion(dynamic_cast<BoidsState*>(neighborbirds) );
    }
   
   	// Arbitrarily weight these forces
    sep *= 1.5;
    ali *= 1.0; // Might need to alter weights for different characteristics
    coh *= 1.0;
    
	// Add the force vectors to acceleration
    applyForce(sep);
    applyForce(ali);
    applyForce(coh);
}

// Adds force to current acceleration
void BirdShm::applyForce(const Vector3D& force)
{
	acceleration = acceleration + force;
}

// Separation
// Keeps boids from getting too close to one another
Vector3D BirdShm::Separation(const BoidsState* boids)
{
    // Distance of field of vision for separation between boids
    Vector3D steer(0, 0, 0);
    int count = 0;
    // For every boid in the system, check if it's too close
    for ( auto& iter : boids->databuffer ) 
	{
        // Calculate distance from current boid to boid we're looking at
		auto diff = iter.second.getPosition() - position;
		auto d = diff.length();
		// If this is a fellow boid and it's too close, move away from it
        if ((d > 0) && (d < seperate_range)) 
		{
            diff.normalize();
            diff /= d;      // Weight by distance
            steer += diff;
            count++;
		}
    }
    // Adds average difference of location to acceleration
    if (count > 0)
        steer /= count;

    if (steer.length() > 0) {
        // Steering = Desired - Velocity
        steer.normalize();
		steer *= maxSpeed;
		steer -= velocity;
		steer.limit(maxForce);
    }

    return steer;
}

// Alignment
// Calculates the average velocity of boids in the field of vision and
// manipulates the velocity of the current bird in order to match it
Vector3D BirdShm::Alignment(const BoidsState* boids)
{

    Vector3D sum(0, 0, 0);
    int count = 0;
    for (auto& iter : boids->databuffer) 
	{
        float d = (position - iter.second.getPosition()).length();
        if ((d > 0) && (d < alignment_range)) // 0 < d < 50
		{
			sum += iter.second.getVelocity();
            count++;
        }
    }
    // If there are boids close enough for alignment...
    if (count > 0) 
	{
		sum /= count;			// Divide sum by the number of close boids (average of velocity)
		sum.normalize();		// Turn sum into a unit vector, and
		sum *= maxSpeed;		// Multiply by maxSpeed

        // Steer = Desired - Velocity
        Vector3D steer;
        steer = sum - velocity;				//sum = desired(average)
		steer.limit(maxForce);
        return steer;

    } else {
		Vector3D tmp;
        return tmp;
    }
}

// Cohesion
// Finds the average location of nearby boids and manipulates the
// steering force to move in that direction.
Vector3D BirdShm::Cohesion(const BoidsState* boids)
{
    
    Vector3D sum(0, 0, 0);
    int count = 0;
    for (auto& iter : boids->databuffer) 
	{
        float d = ( position - iter.second.getPosition() ).length();
        if ((d > 0) && (d < cohesion_range)) 
		{
			sum += iter.second.getPosition();
            count++;
        }
    }
    if (count > 0) 
	{
		sum /= count;
		return sum;

        //sum.divScalar(count);
        //return std::move( seek(sum) );
    } else 
	{
        Vector3D temp(0, 0, 0);
        return temp;
    }
}
/*
// Limits the maxSpeed, finds necessary steering force and
// normalizes vectors
Vector3D& BirdShm::seek(const Vector3D& v)
{
    Vector3D* desired = new Vector3D();
    desired.subVector(v);  // A vector pointing from the location to the target
    // Normalize desired and scale to maximum speed
    desired.normalize();
    desired.mulScalar(maxSpeed);
    // Steering = Desired minus Velocity
    acceleration.subTwoVector(desired, velocity);
    acceleration.limit(maxForce);  // Limit to maximum steering force
    return acceleration;
}

*/

File: modeling_on_python/examples/boids_shm/bird_shm.py
"""
bird_shm.py - Boids鸟群仿真实体

实现了经典的Boids算法三大规则：
1. 分离(Separation) - 避免过近的邻居
2. 对齐(Alignment) - 与邻居保持相同方向
3. 聚合(Cohesion) - 向邻居的中心移动
"""

import sys
import os
import random
import math

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../python'))

try:
    import causal
except ImportError:
    print("错误：无法导入causal模块")
    sys.exit(1)

from vector3d import Vector3D
from message import BirdTrace, BoidsState, T_SelfWork, T_BirdTrace


class BirdShm(causal.SimEntity):
    """
    Boids鸟群仿真实体
    
    每只鸟遵循三个规则：分离、对齐、聚合
    """

    def __init__(self, position, velocity, separate_range=20.0, cohesion_range=15.0):
        """
        初始化鸟实体
        
        参数:
            position (Vector3D): 初始位置
            velocity (Vector3D): 初始速度
            separate_range (float): 分离规则的作用距离
            cohesion_range (float): 聚合规则的作用距离
        """
        super().__init__()
        self.position = position.copy()
        self.velocity = velocity.copy()
        self.acceleration = Vector3D(0, 0, 0)

        # 参数
        self.max_speed = 4.0
        self.max_force = 0.1
        self.separate_range = separate_range
        self.alignment_range = separate_range * 1.5  # 对齐范围稍大
        self.cohesion_range = cohesion_range

        self.delta = 0.1  # 时间步长

    def Init(self):
        """初始化实体，发送第一个自我工作消息"""
        # 将初始轨迹添加到共享状态
        boids_state = self.getIntraLPState()
        if boids_state:
            trace = BirdTrace(self.EntityID(), self.position, self.velocity)
            boids_state.add(self.EntityID(), trace)

        # 发送第一个自我工作消息
        msg = causal.SimMsg(T_SelfWork)
        self.send(self.EntityID(), msg, causal.SimTime(self.delta))

    def execute(self, msg):
        """
        处理接收到的消息
        
        参数:
            msg (SimMsg): 接收到的消息
        """
        msg_type = msg.getMsgType()

        if msg_type == T_SelfWork:
            self.cycle_work()
        else:
            print(f"Warning: Unknown message type {msg_type}")

    def Terminate(self, ts):
        """终止实体"""
        pass

    def cycle_work(self):
        """
        周期性工作：
        1. 从共享状态读取邻居信息
        2. 应用鸟群规则
        3. 更新位置和速度
        4. 将轨迹写入共享状态
        5. 发送下一个工作消息
        """
        # 1. 从共享状态读取邻居信息
        boids_state = self.getIntraLPState()

        # 2. 应用鸟群规则
        self.flock(boids_state)

        # 3. 更新速度和位置
        self.velocity += self.acceleration
        self.velocity.limit(self.max_speed)
        self.position += self.velocity * self.delta

        # 边界处理：环绕
        self.wrap_around()

        # 重置加速度
        self.acceleration = Vector3D(0, 0, 0)

        # 4. 将轨迹写入共享状态
        trace = BirdTrace(self.EntityID(), self.position, self.velocity)
        boids_state.modify(self.EntityID(), trace)

        # 5. 发送下一个工作消息
        msg = causal.SimMsg(T_SelfWork)
        self.send(self.EntityID(), msg, causal.SimTime(self.delta))

    def flock(self, boids_state):
        """
        应用鸟群规则
        
        参数:
            boids_state (BoidsState): 共享状态对象
        """
        # 计算三个力
        sep = self.separation(boids_state)
        ali = self.alignment(boids_state)
        coh = self.cohesion(boids_state)

        # 权重系数
        sep *= 1.5
        ali *= 1.0
        coh *= 1.0

        # 应用力
        self.apply_force(sep)
        self.apply_force(ali)
        self.apply_force(coh)

    def apply_force(self, force):
        """
        应用力到加速度
        
        参数:
            force (Vector3D): 力向量
        """
        self.acceleration += force

    def separation(self, boids_state):
        """
        分离规则：避免过近的邻居
        
        参数:
            boids_state (BoidsState): 共享状态对象
            
        返回:
            Vector3D: 分离力向量
        """
        steer = Vector3D(0, 0, 0)
        count = 0

        # 遍历所有鸟
        for bird_id, trace in boids_state.databuffer.items():
            if bird_id == self.EntityID():
                continue

            d = self.position.distance(trace.position)

            # 如果在分离范围内
            if 0 < d < self.separate_range:
                # 计算远离的方向
                diff = self.position - trace.position
                diff.normalize()
                diff /= d  # 距离越近，力越大
                steer += diff
                count += 1

        # 平均
        if count > 0:
            steer /= count

        # 如果有力，实现steering = desired - velocity
        if steer.length() > 0:
            steer.normalize()
            steer *= self.max_speed
            steer -= self.velocity
            steer.limit(self.max_force)

        return steer

    def alignment(self, boids_state):
        """
        对齐规则：与邻居保持相同方向
        
        参数:
            boids_state (BoidsState): 共享状态对象
            
        返回:
            Vector3D: 对齐力向量
        """
        sum_vel = Vector3D(0, 0, 0)
        count = 0

        # 遍历所有鸟
        for bird_id, trace in boids_state.databuffer.items():
            if bird_id == self.EntityID():
                continue

            d = self.position.distance(trace.position)

            # 如果在对齐范围内
            if 0 < d < self.alignment_range:
                sum_vel += trace.velocity
                count += 1

        if count > 0:
            sum_vel /= count
            sum_vel.normalize()
            sum_vel *= self.max_speed

            steer = sum_vel - self.velocity
            steer.limit(self.max_force)
            return steer
        else:
            return Vector3D(0, 0, 0)

    def cohesion(self, boids_state):
        """
        聚合规则：向邻居的中心移动
        
        参数:
            boids_state (BoidsState): 共享状态对象
            
        返回:
            Vector3D: 聚合力向量
        """
        sum_pos = Vector3D(0, 0, 0)
        count = 0

        # 遍历所有鸟
        for bird_id, trace in boids_state.databuffer.items():
            if bird_id == self.EntityID():
                continue

            d = self.position.distance(trace.position)

            # 如果在聚合范围内
            if 0 < d < self.cohesion_range:
                sum_pos += trace.position
                count += 1

        if count > 0:
            sum_pos /= count
            return self.seek(sum_pos)
        else:
            return Vector3D(0, 0, 0)

    def seek(self, target):
        """
        寻找目标位置
        
        参数:
            target (Vector3D): 目标位置
            
        返回:
            Vector3D: 导向力向量
        """
        desired = target - self.position
        desired.normalize()
        desired *= self.max_speed

        steer = desired - self.velocity
        steer.limit(self.max_force)
        return steer

    def wrap_around(self):
        """边界处理：环绕到另一边"""
        # 这些值应该从配置文件读取，这里使用默认值
        length = 100.0
        width = 100.0
        height = 100.0

        if self.position.x < 0:
            self.position.x = length
        if self.position.x > length:
            self.position.x = 0

        if self.position.y < 0:
            self.position.y = width
        if self.position.y > width:
            self.position.y = 0

        if self.position.z < 0:
            self.position.z = height
        if self.position.z > height:
            self.position.z = 0

    def get_position(self):
        """获取当前位置"""
        return self.position
